lpc55_sign/src/debug_auth.rs

```diff
@@ -0,0 +1,213 @@
+use lpc55_areas::{CCSOCUDflt, CCSOCUPin, DebugSettings};
+use num_traits::ToPrimitive;
+use rsa::{PublicKeyParts, RsaPrivateKey, RsaPublicKey};
+use sha2::{Digest, Sha256};
+use x509_cert::Certificate;
+
+use crate::{
+    cert::public_key,
+    signed_image::{pad_roots, root_key_hash},
+    Error,
+};
+
+const SOCC: u32 = 0x0000_0001;
+
+#[derive(Debug)]
+pub struct DebugAuthChallenge {
+    // NXP UM11126 claims this is a single little-endian u32 version field.
+    // NXP's spsdk tooling implements it as done here: two little-endian u16s
+    // with version_major at the lower offset.  See
+    // https://github.com/nxp-mcuxpresso/spsdk/blob/5da31d96a020bd65e5834ea5ac1b68327ea965ef/spsdk/dat/dac_packet.py#L114.
+    version_major: u16,
+    version_minor: u16,
+
+    socc: u32,
+    uuid: [u8; 16],
+
+    rotk_revoke: [bool; 4],
+    rotkh: [u8; 32],
+
+    cc_socu_pin: CCSOCUPin,
+    cc_socu_dflt: CCSOCUDflt,
+    vendor_usage: u32,
+
+    challenge_vector: [u8; 32],
+}
+
+#[derive(thiserror::Error, Debug)]
+pub enum DebugAuthChallengeParseError {
+    #[error("Input wrong length (size={0}, expected=104)")]
+    WrongLength(usize),
+}
+
+impl TryFrom<&[u8]> for DebugAuthChallenge {
```

Comment:
We could use `derive(zerocopy::FromBytes)` on the `struct DebugAuthChallenge`, along with explicit endianness, e.g. `zerocopy::U16<LittleEndian>`, to remove most of this implementation.

Two other changes would be required:
- `CCSOCUPin` and `CCSOCUDflt` would also need to be annotated with `#[derive(FromBytes)]`
- `rotk_revoke` would have to be `[u8; 4]`, since `bool` isn't `FromBytes`.

This could be more robust than hand-rolling the deserialization code:
```rust
impl TryFrom<&[u8]> for DebugAuthChallenge {
    type Error = DebugAuthChallengeParseError;

    fn try_from(value: &[u8]) -> Result<Self, Self::Error> {
        use zerocopy::FromBytes;
        Self::read_from(value).ok_or(Self::Error::WrongLength(value.len()))
    }
}
```

Reply:
(You'd also want to `repr(C)` it -- as a reader I was surprised it wasn't `repr(C)` until I saw that we're hand-parsing it.)

Reply:
It looks like `rotk_revoke` would need to be e.g. a `bitflags` type rather than `[u8; 4]` to match the decoding code below, fwiw.